ghga_datasteward_kit/s3_upload/entrypoint.py

```diff
@@ -0,0 +1,230 @@
+
+"""
+Contains functionality to actually run the S3 upload.
+"""
+
+import asyncio
+import base64
+import logging
+from pathlib import Path
+
+import typer
+from ghga_service_commons.utils.crypt import encrypt
+
+from ghga_datasteward_kit import models
+from ghga_datasteward_kit.s3_upload.config import Config, LegacyConfig
+from ghga_datasteward_kit.s3_upload.downloader import ChunkedDownloader
+from ghga_datasteward_kit.s3_upload.uploader import ChunkedUploader
+from ghga_datasteward_kit.s3_upload.utils import (
+    LOGGER,
+    HttpxClientState,
+    check_adjust_part_size,
+    check_output_path,
+    handle_superficial_error,
+    httpx_client,
+    objectstorage,
+)
+from ghga_datasteward_kit.utils import load_config_yaml, read_token
+
+
+async def shared_path(input_path: Path, alias: str, config: LegacyConfig):
+    """
+    Functionality shared between legacy and non-legacy upload path
+    """
+    if not input_path.exists():
+        msg = f"No such file: {input_path.resolve()}"
+        handle_superficial_error(msg=msg)
+
+    if input_path.is_dir():
+        msg = f"File location points to a directory: {input_path.resolve()}"
+        handle_superficial_error(msg=msg)
+
+    check_output_path(config.output_dir / f"{alias}.json")
+
+    file_size = input_path.stat().st_size
+    check_adjust_part_size(config=config, file_size=file_size)
+
+    # set retry policy
+    HttpxClientState.configure(5)
+
+    uploader = ChunkedUploader(
+        input_path=input_path,
+        alias=alias,
+        config=config,
+        unencrypted_file_size=file_size,
+    )
+    await uploader.encrypt_and_upload()
+
+    downloader = ChunkedDownloader(
+        config=config,
+        file_id=uploader.file_id,
+        encrypted_file_size=uploader.encryptor.encrypted_file_size,
+        file_secret=uploader.encryptor.file_secret,
+        part_size=config.part_size,
+        target_checksums=uploader.encryptor.checksums,
+    )
+    await downloader.download()
+
+    return uploader, file_size
+
+
+async def exchange_secret_for_id(
+    *, file_id: str, alias: str, secret: bytes, token: str, config: Config
+) -> str:
+    """
+    Call file ingest service to store the file secret and obtain a secret ID by which
+    it can be retrieved.
+
+    If storing the secret fails, the uploaded file is deleted from object storage and
+    a ValueError is raised containing the file alias and response status code.
+    """
+
+    file_secret = base64.b64encode(secret).decode("utf-8")
+    payload = encrypt(data=file_secret, key=config.secret_ingest_pubkey)
+    encrypted_secret = models.EncryptedPayload(payload=payload)
+
+    with httpx_client() as client:
+        headers = {"Authorization": f"Bearer {token}"}
+        response = client.post(
+            url=config.secret_ingest_url,
+            json=encrypted_secret.dict(),
+            headers=headers,
+            timeout=60,
+        )
+
+        if response.status_code != 200:
```

Comment:
I wouldn't see the cleanup of the bucket as the responsibility of this function. However, if you cannot find an easy alternative, this is also fine.

Reply:
See next comment